src/events/base.py

```diff
@@ -68,6 +70,13 @@ def get_app_status(  # noqa: C901
         if self.charm.app.planned_units() > 1 and not self.context.zookeeper:
             return Status.MISSING_ZOOKEEPER.value
 
+        if self.charm.config.get("expose-external", "false") not in VALID_EXPOSE_EXTERNAL_VALUES:
+            return Status.INVALID_EXPOSE_EXTERNAL.value
```

Comment:
I've removed this validation from here, and have also implemented `structed_config` along this PR.

One thing I noticed with the `structured_config` and thus subsequently the `data_models` lib is that the lib doesn't set the charm to "Blocked" state by default when the config validation fails. And since the `pydantic`'s `ValidationError` is just raised by the charm thus crashing the charm whenever `charm.config` is accessed, the `try / except` and `set_block_status()` should be done at the earliest level of `charm.py` itself and not in the `compute_status` decorator as we have desired for central place of setting app status.

For the scope of this PR, I have left the validation entirely to `structured_config` and thus the charm will crash when `expose-external` is not one of permitted values. As @welpaolo and me discussed, this would not be much of an issue at the moment, but I'd like to have your opinions as well, @deusebio.

Reply:
I would indeed put a try_catch logic in the constructors of the charm and if uncaught validation errors occurs, I would set the status into blocked status and return straight away (such that hooks are not even registered and run)

Reply:
Indeed adding be a `try..except` logic in the charm constructor is the most straightforward workaround. However, this will be against our preference of having a single central place of setting charm statust the es. Even if we go along with this as a temporary solution, I'd prefer we look into a more robust solution if possible inside the `data_models` library itself, such that it can set the charm in BlockedState when a ValidationError occurs.

Reply:
> I think this might cause unexpected issues. By letting the charm error, we make sure that juju will queue events before trying again, whereas an early return before registering events hook handlers would drop events fired while the configuration was wrong.
Maybe sys.exit(1) would work instead?

@Batalex I too experienced this exact issue. `sys.exit(1)` would cause the charm to error. However, exiting from the constructor itself would effectively drop all events fired when the configuration is wrong. The app status was set to desired blocked state, but the unit always had status "error" with message `hook failed: <hook-name>`. In addition to this, the charm also repeatedly tries to resolve the error (see screenshot below). I'm not sure this is what we want. cc: @deusebio @welpaolo 

![image](https://github.com/user-attachments/assets/c1a288d3-7a09-4cb1-8b59-2fdbc9889069)

Reply:
In addition to this, once the charm went to error, `juju config kyuubi-k8s expose-external=false` did not bring it back. The logs showed `juju.worker.uniter awaiting error resolution for "config-changed" hook` and I had to do `juju resolve kyuubi-k8s/0` and wait for a long time to have the unit recover.

Reply:
No I don't think that this is what we want. The `sys.exit(1)` strategy does not sound really good. I would prefer to go with the same approach that Kafka and Zookeeper is using.

Reply:
I generally don't like charms to go into error state, which feels an uncaught and handled exception. Also, when the config is wrong, my expected ux would be to block and ask the user to change the value, not error out IMHO. 

Having said this, that's a very fair point that my naive suggestion of returning logic in the constructor is arguably even worse, since we are skipping events rather than deferring.

Anyhow, let's not address this here, and don't over engineer this. I agree to keep things with validation error that may be raised, which is the simplest solution from the code perspective and has a resolution path (using juju resolve).

I would still try to have a thought how this can be improved. I will try to ask another couple of teams that I know they are using pydantic